valueset.go

```diff
@@ -0,0 +1,247 @@
+// 	for _, nsD := range nodeset.Data {
+// 		nodes = append(nodes, grph.NewNode(nsD.Labels, nsD.Properties))
+// 	}
+// 	return nodes
+// }
+
+// oldNodeset is nodeset pulled from DB
+func diff(oldNodeset, newNodeset Nodeset) (insertions NodesetData, deletions []string, updates NodesetData) {
+	// compare roots
+	if oldNodeset.ID != newNodeset.ID {
+		// update
+		updates.ID = newNodeset.ID
+		if !reflect.DeepEqual(oldNodeset.Labels, newNodeset.Labels) {
+			// update
+			updates.Labels = newNodeset.Labels
+		}
+	}
+	seenOldNodesetData := make(map[*NodesetData]struct{})
+	for _, oldNsData := range oldNodeset.Data {
+		for _, newNsData := range newNodeset.Data {
+			// if newNodeset is not in oldNodeset (newNodeset must === oldNodeset)
+			if _, has := seenOldNodesetData[&newNsData]; !has {
+				// delete
+				deletions = append(deletions, oldNsData.ID)
+			}
+			for k, v := range newNsData.Properties {
+				// if oldNodeset does not have property
+				if _, has := oldNsData.Properties[k]; !has {
+					// insert
+					insertions.Properties[k] = v
+				} else {
+					// compare, nop or update
+					if !reflect.DeepEqual(oldNsData.Properties, newNsData.Properties) {
+						// update
+						updates.Properties[k] = v
+					}
+				}
+			}
+			if !lpg.NewStringSet(oldNsData.Labels...).IsEqual(lpg.NewStringSet(newNsData.Labels...)) {
+				// update - setting labels to those in newNodeset but not in oldNodeset
+				updates.Labels = findLabelDiff(lpg.NewStringSet(newNsData.Labels...), (lpg.NewStringSet(oldNsData.Labels...)))
+			}
+			seenOldNodesetData[&oldNsData] = struct{}{}
+		}
+	}
+	return insertions, deletions, updates
+}
+
+/*
+where insertions and deletions are node ids. (NodesetData.ID)
+Find NodesetData that are in oldNodeset but not in NewNodeset -> deletions
+Find NodesetData that are in newNodeset but not in oldNodeset: -> insertions
+Find NodesetData that are in both oldNodeset and NewNodeset: make sure labels and props are same, if not -> updates
+
+Also write:
+*/
+func LoadNodeset(tx neo4j.Transaction, nodesetId string) (Nodeset, error) {
+	query := "MATCH (root:`NODESET` {nodeset_id: $id})-[:$id]->(m) return root,m"
+	idrec, err := tx.Run(query, map[string]interface{}{"id": nodesetId})
+	if err != nil {
+		return Nodeset{}, err
+	}
+	ns := Nodeset{
+		ID:     nodesetId,
+		Labels: make([]string, 0),
+		Data:   make(map[string]NodesetData),
+	}
+	first := false
+	for idrec.Next() {
+		rec := idrec.Record()
+		if !first {
+			nsCenter := rec.Values[0].(neo4j.Node)
+			ns.Labels = nsCenter.Labels
+			first = true
+		}
+		nsLeaf := rec.Values[1].(neo4j.Node)
+		var nsLeafEntityId string
+		for k := range nsLeaf.Props {
+			if k == ls.EntityIDTerm {
+				nsLeafEntityId = k
+			}
+		}
+		ns.Data[nsLeafEntityId] = NodesetData{
+			ID:         nsLeafEntityId,
+			Labels:     nsLeaf.Labels,
+			Properties: nsLeaf.Props,
+		}
+	}
+	return ns, nil
+}
+
+func (ns Nodeset) Execute(tx neo4j.Transaction, op string) error {
```

Comment:
Execute can get only inserts, updates, deletes, and what to do with the root node.

If op=apply: Load from db, and compute diff
If op=delete: Load from db, diff({},dbNs)

It should be:
LoadNodes
Diff
Execute